### PR TITLE
nsqd: enable block profile

### PR DESCRIPTION
**Pull request by absolute8511**

### CODE REVIEW

**Comment by mreiferson** [comment]
@absolute8511 this is already available at runtime via the HTTP pprof endpoints http://nsq.io/components/nsqd.html#get-debugpprofblock, closing.

**Comment by absolute8511** [comment]
block profile will not be enabled by default if you did not `SetBlockProfileRate`

**Comment by mreiferson** [comment]
@absolute8511 sorry! didn't realize that - it seems like this is something you don't want on all the time (hence the default being off).  Perhaps we need another `/debug/...` endpoint to temporarily enable it?

**Comment by absolute8511** [comment]
Maybe using `/debug/...` would be better, but I am not sure is it safe to set this dynamically.

**Comment by mreiferson** [comment]
> Maybe using /debug/... would be better, but I am not sure is it safe to set this dynamically.

According to http://grokbase.com/p/gg/golang-nuts/152ra4r1vz/go-nuts-trying-to-use-the-pprof-block-endpoint, that's the way it's expected to be used.

File: nsqd/http.go [hunk 1]
```diff
@@ -99,11 +100,23 @@ func newHTTPServer(ctx *context, tlsEnabled bool, tlsRequired bool) *httpServer
 	router.Handler("GET", "/debug/pprof/heap", pprof.Handler("heap"))
 	router.Handler("GET", "/debug/pprof/goroutine", pprof.Handler("goroutine"))
 	router.Handler("GET", "/debug/pprof/block", pprof.Handler("block"))
+	router.HandlerFunc("PUT", "/debug/blockrate",
```
Comment (mreiferson) [inline nsqd/http.go]: let's name this `/debug/setblockrate` and keep this on one line

Comment (mreiferson) [inline nsqd/http.go]: also, let's log this, something like:

```go
router.Handle("PUT", "/debug/setblockrate", http_api.Decorate(s.setBlockRateHandler, log, http_api.PlainText))
```

File: nsqd/http.go [hunk 2]
```diff
@@ -99,11 +100,23 @@ func newHTTPServer(ctx *context, tlsEnabled bool, tlsRequired bool) *httpServer
...
 	router.Handler("GET", "/debug/pprof/threadcreate", pprof.Handler("threadcreate"))
 
 	return s
 }
 
+func HandleBlockRate(w http.ResponseWriter, req *http.Request) {
+	rate, err := strconv.Atoi(req.FormValue("rate"))
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		fmt.Fprintf(w, "block rate value error: %s", err.Error())
```
Comment (mreiferson) [inline nsqd/http.go]: after this is converted to the custom HTTP handler, we can just:

```go
return nil, http_api.Err{400, fmt.Sprintf("invalid block rate %s", err)}
```

File: nsqd/http.go [hunk 3]
```diff
@@ -99,11 +100,23 @@ func newHTTPServer(ctx *context, tlsEnabled bool, tlsRequired bool) *httpServer
...
+		return
+	}
+	runtime.SetBlockProfileRate(rate)
```
Comment (mreiferson) [inline nsqd/http.go]: should we return output in the success case?

Comment (absolute8511) [inline nsqd/http.go]: Since the SetBlockProfileRate has no return value, I think we no need to return output.

File: nsqd/http.go [hunk 4]
```diff
@@ -99,11 +100,21 @@ func newHTTPServer(ctx *context, tlsEnabled bool, tlsRequired bool) *httpServer
 	router.Handler("GET", "/debug/pprof/heap", pprof.Handler("heap"))
 	router.Handler("GET", "/debug/pprof/goroutine", pprof.Handler("goroutine"))
 	router.Handler("GET", "/debug/pprof/block", pprof.Handler("block"))
+	router.Handle("PUT", "/debug/setblockrate", http_api.Decorate(HandleBlockRate, log, http_api.PlainText))
 	router.Handler("GET", "/debug/pprof/threadcreate", pprof.Handler("threadcreate"))
 
 	return s
 }
 
+func HandleBlockRate(w http.ResponseWriter, req *http.Request, ps httprouter.Params) (interface{}, error) {
```
Comment (mreiferson) [inline nsqd/http.go]: Last comment!  This doesn't need to be exported, let's name it `setBlockRateHandler` please

**Comment by mreiferson** [comment]
@absolute8511 thanks for this - after that last naming change can you squash down to one commit please?

**Comment by absolute8511** [comment]
@mreiferson done.

**Comment by mreiferson** [comment]
Thanks!

**Pull request merged**